File: languages/js/src/entity/record/record.ts
import { BloockBridge } from "../../bridge/bridge";
import * as proto from "../../bridge/proto/record_entities";
import {
  GetHashRequest,
  GetPayloadRequest,
  SetProofRequest
} from "../../bridge/proto/record";
import { ConfigData } from "../../bridge/proto/config";
import { Proof } from "../integrity";

/**
 * Represents a record with payload, hash, and configuration data.
 */
export class Record {
  payload: Uint8Array;
  hash: string;
  configData: ConfigData;

  /**
   * Constructs a Record object with the specified parameters.
   * @param payload 
   * @param hash 
   * @param configData 
   */
  constructor(payload: Uint8Array, hash: string, configData: ConfigData) {
    this.payload = payload;
    this.hash = hash;
    this.configData = configData;
  }

  static fromProto(r: proto.Record, configData: ConfigData) {
    return new Record(r.payload, r.hash, configData);
  }

  toProto(): proto.Record {
    return proto.Record.fromPartial({
      configData: this.configData,
      payload: this.payload,
      hash: this.hash
    });
  }

  /**
   * Retrieves the hash of the record.
   * @returns 
   */
  async getHash(): Promise<string> {
    const bridge = new BloockBridge();
    return bridge
      .getRecord()
      .GetHash(
        GetHashRequest.fromPartial({
          configData: this.configData,
          record: this.toProto()
        })
      )
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return res.hash;
      });
  }

  /**
   * Retrieves the payload of the record.
   * @returns 
   */
  async getPayload(): Promise<Uint8Array> {
    const bridge = new BloockBridge();
    return bridge
      .getRecord()
      .GetPayload(
        GetPayloadRequest.fromPartial({
          configData: this.configData,
          record: this.toProto()
        })
      )
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return res.payload;
      });
  }

  /**
   * Returns the payload of the record.
   * @returns 
   */
  public retrieve(): Uint8Array {
    return this.payload;
  }

  /**
   * Sets the proof for a record.
   * @param proof 
   * @returns 
   */
  async setProof(proof: Proof) {
    const bridge = new BloockBridge();

    const req = SetProofRequest.fromPartial({
      configData: this.configData,
      record: this.toProto(),
      proof: proof.toProto()
    });

    return bridge
      .getRecord()
      .SetProof(req)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        this.payload = res.record?.payload!;
      });
  }
}

export class RecordHeader {
  ty: string;
  constructor(ty: string) {
    this.ty = ty;
  }

  static fromProto(recordHeader: proto.RecordHeader): RecordHeader {
    return new RecordHeader(recordHeader.ty);
  }

  toProto(): proto.RecordHeader {
    return proto.RecordHeader.fromPartial({ ty: this.ty });
  }
}

File: languages/go/entity/record/record.go
package record

import (
	"context"
	"errors"

	"github.com/bloock/bloock-sdk-go/v2/entity/integrity"
	"github.com/bloock/bloock-sdk-go/v2/internal/bridge"
	"github.com/bloock/bloock-sdk-go/v2/internal/bridge/proto"
)

// Record represents a record with payload, hash, and configuration data.
type Record struct {
	Payload    []byte
	hash       string
	configData *proto.ConfigData
}

func NewRecordFromProto(r *proto.Record, configData *proto.ConfigData) Record {
	if r == nil {
		return Record{}
	}

	return Record{
		Payload:    r.Payload,
		hash:       r.Hash,
		configData: configData,
	}
}

func (r *Record) ToProto() *proto.Record {
	return &proto.Record{
		ConfigData: r.configData,
		Payload:    r.Payload,
		Hash:       r.hash,
	}
}

// GetHash retrieves the hash of the record.
func (r *Record) GetHash() (string, error) {
	bridgeClient := bridge.NewBloockBridge()
	res, err := bridgeClient.Record().GetHash(context.Background(), &proto.GetHashRequest{ConfigData: r.configData, Record: r.ToProto()})

	if err != nil {
		return "", err
	}

	if res.Error != nil {
		return "", errors.New(res.Error.Message)
	}

	return res.Hash, nil
}

// GetPayload retrieves the payload of the record.
func (r *Record) GetPayload() ([]byte, error) {
	bridgeClient := bridge.NewBloockBridge()
	res, err := bridgeClient.Record().GetPayload(context.Background(), &proto.GetPayloadRequest{ConfigData: r.configData, Record: r.ToProto()})

	if err != nil {
		return nil, err
	}

	if res.Error != nil {
		return nil, errors.New(res.Error.Message)
	}

	return res.Payload, nil
}

// Retrieve returns the payload of the record.
func (r *Record) Retrieve() []byte {
	return r.Payload
}

// SetProof sets the proof for a record.
func (r *Record) SetProof(proof integrity.Proof) error {
	bridgeClient := bridge.NewBloockBridge()
	res, err := bridgeClient.Record().SetProof(context.Background(), &proto.SetProofRequest{
		ConfigData: r.configData,
		Record:     r.ToProto(),
		Proof:      proof.ToProto(),
	})

	if err != nil {
		return err
	}

	if res.Error != nil {
		return errors.New(res.Error.Message)
	}

	r.Payload = res.Record.Payload

	return nil
}

func MapRecordsToProto(records []Record) []*proto.Record {
	recordsProto := make([]*proto.Record, len(records))
	for i, record := range records {
		recordsProto[i] = record.ToProto()
	}
	return recordsProto
}
